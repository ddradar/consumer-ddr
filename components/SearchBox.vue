<template>
  <section>
    <OLoading :active="isLoading" />
    <OAutocomplete
      v-model="term"
      icon="magnify"
      placeholder="曲名 or アーティスト名"
      clearable
      :data="songs"
      :loading="isLoading"
      @select="onSelected"
    >
      <template #default="props">
        <span>{{ props.option.name }} / {{ props.option.artist }}</span>
      </template>
    </OAutocomplete>
  </section>
</template>

<script lang="ts" setup>
import type { Song } from '~~/src/content'

const keys = ['slug', 'name', 'artist'] as const
type SearchSongResult = Pick<Song, typeof keys[number]>

const term = ref('')

const { data: _songs, pending: isLoading } = await useLazyAsyncData(
  '/search-box',
  () =>
    queryContent<Song>()
      .where({ _type: 'json' })
      .only([...keys])
      .find(),
  {
    transform: (songs) =>
      songs.reduce((prev, s) => {
        if (!prev.find((d) => d.slug === s.slug)) prev.push(s)
        return prev
      }, [] as SearchSongResult[])
  }
)
const songs = computed(() =>
  term.value
    ? _songs.value?.filter(
        (d) =>
          d.name.toUpperCase().includes(term.value.toUpperCase()) ||
          d.artist.toUpperCase().includes(term.value.toUpperCase())
      )
    : []
)

const router = useRouter()
const onSelected = (song: SearchSongResult) => {
  term.value = song.name
  router.push(`/songs/${song.slug}`)
}
</script>
