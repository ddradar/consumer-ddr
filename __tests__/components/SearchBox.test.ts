import { describe, expect, test, vi } from 'vitest'
import { ref } from 'vue'

import SearchBox, { SearchSongResult } from '~~/components/SearchBox.vue'

import { mountAsync, plugins } from '../test-utils'

describe('components/SearchBox.vue', () => {
  const songData: SearchSongResult[] = [
    { slug: 'boys-smiledk', name: 'BOYS', artist: 'smile. dk' },
    { slug: 'butterfly', name: 'butterfly', artist: 'smile. dk' },
    { slug: 'little-bitch', name: 'LITTLE BITCH', artist: 'THE SPECIALS' },
    { slug: 'my-fire', name: 'MY FIRE', artist: 'X-TREME' }
  ]

  describe('snapshot test', () => {
    test('{ isLoading: true, term: "" } renders loading state', async () => {
      // Arrange
      const data = ref([])
      const pending = ref(true)
      vi.mocked(useLazyAsyncData).mockResolvedValue({ data, pending } as any)

      // Act
      const wrapper = await mountAsync(SearchBox, { global: { plugins } })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
    test('{ isLoading: false, term: "" } renders input without autocomplete', async () => {
      // Arrange
      const data = ref([])
      const pending = ref(false)
      vi.mocked(useLazyAsyncData).mockResolvedValue({ data, pending } as any)

      // Act
      const wrapper = await mountAsync(SearchBox, { global: { plugins } })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('term', () => {
    test('@change renders autocomplete', async () => {
      // Arrange
      const data = ref(songData)
      const pending = ref(false)
      vi.mocked(useLazyAsyncData).mockResolvedValue({ data, pending } as any)

      // Act
      const wrapper = await mountAsync(SearchBox, { global: { plugins } })
      await wrapper.find('input.input').setValue('sm')

      // Assert
      expect(wrapper.findAll('.dropdown-item')).toHaveLength(2)
    })
  })

  describe('autocomplete', () => {
    test('@select calls router.push("/songs/{id}")', async () => {
      // Arrange
      const data = ref(songData)
      const pending = ref(false)
      vi.mocked(useLazyAsyncData).mockResolvedValue({ data, pending } as any)
      const push = vi.fn()
      vi.mocked(useRouter).mockReturnValue({ push } as any)

      // Act
      const wrapper = await mountAsync(SearchBox, { global: { plugins } })
      await wrapper.find('input.input').setValue('sm')
      await wrapper.find('.dropdown-item').trigger('click')

      // Assert
      expect(push).toBeCalledWith(`/songs/${songData[0].slug}`)
    })
  })
})
