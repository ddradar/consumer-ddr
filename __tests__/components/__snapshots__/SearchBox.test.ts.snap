// Vitest Snapshot v1

exports[`components/SearchBox.vue > snapshot test > { isLoading: false, term: "" } renders input without autocomplete 1`] = `
<section>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <div
    class="autocomplete control"
  >
    <div
      class="control has-icons-left"
    >
      <input
        aria-autocomplete="list"
        autocomplete="off"
        class="input"
        loading="false"
        placeholder="曲名 or アーティスト名"
        type="text"
      />
      <span
        class="icon is-left"
      >
        <i
          class="mdi mdi-magnify mdi-24px"
        />
      </span>
      <!--v-if-->
      <!--v-if-->
    </div>
    <transition-stub>
      <div
        class="dropdown-menu dropdown-content is-opened-bottom"
        is="div"
        style="display: none;"
      >
        <!--v-if-->
        
        
        <!--v-if-->
        
        
        
        
        <!--v-if-->
        <!--v-if-->
      </div>
    </transition-stub>
  </div>
</section>
`;

exports[`components/SearchBox.vue > snapshot test > { isLoading: true, term: "" } renders loading state 1`] = `
<section>
  <transition-stub>
    <div
      class="o-load loading o-load--fullpage is-full-page"
    >
      <div
        class="o-load__overlay loading-overlay"
      />
      
      <span
        class="icon is-spin is-medium o-load__icon icon"
      >
        <i
          class="mdi mdi-loading mdi-36px"
        />
      </span>
      
    </div>
  </transition-stub>
  <div
    class="autocomplete control"
  >
    <div
      class="control has-icons-left"
    >
      <input
        aria-autocomplete="list"
        autocomplete="off"
        class="input"
        loading="true"
        placeholder="曲名 or アーティスト名"
        type="text"
      />
      <span
        class="icon is-left"
      >
        <i
          class="mdi mdi-magnify mdi-24px"
        />
      </span>
      <!--v-if-->
      <!--v-if-->
    </div>
    <transition-stub>
      <div
        class="dropdown-menu dropdown-content is-opened-bottom"
        is="div"
        style="display: none;"
      >
        <!--v-if-->
        
        
        <!--v-if-->
        
        
        
        
        <!--v-if-->
        <!--v-if-->
      </div>
    </transition-stub>
  </div>
</section>
`;
